### PR TITLE
BUG: __getitem__ raise blank KeyError for IntervalIndex and missing keys

### LINKED ISSUE #27365: ERR: missing key not displayed in exception message for IntervalIndex
#### Code Sample, a copy-pastable example if possible

adapted from `test_non_matching` in `pandas\tests\indexing\interval\test_interval.py`

```python
>>> import numpy as np
>>> import pandas as pd
>>> pd.__version__
'0.25.0rc0+59.g0437f6899'
>>>
>>> s=pd.Series(np.arange(5), pd.IntervalIndex.from_breaks(np.arange(6)))
>>> s
(0, 1]    0
(1, 2]    1
(2, 3]    2
(3, 4]    3
(4, 5]    4
dtype: int32
>>>
>>> s.loc[[4,5]]
(3, 4]    3
(4, 5]    4
dtype: int32
>>>
>>> s.loc[[4,5,6]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1409, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1816, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1118, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1060, in _get_listlike_indexer
    indexer, keyarr = ax._convert_listlike_indexer(key, kind=self.name)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexes\base.py", line 3239, in _convert_listlike_indexer
    indexer = self._convert_list_indexer(keyarr, kind=kind)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexes\interval.py", line 626, in _convert_list_indexer
    raise KeyError
KeyError
>>>
>>> s.to_frame().loc[[4,5,6]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1409, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1816, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1118, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1060, in _get_listlike_indexer
    indexer, keyarr = ax._convert_listlike_indexer(key, kind=self.name)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexes\base.py", line 3239, in _convert_listlike_indexer
    indexer = self._convert_list_indexer(keyarr, kind=kind)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexes\interval.py", line 626, in _convert_list_indexer
    raise KeyError
KeyError
>>>


```
#### Problem description

With a list indexer, the missing key should be in the exception message
 
#### Expected Output
```python
KeyError: 6
```

another example - from `test_loc_with_overlap` in `pandas\tests\indexing\interval\test_interval_new.py`

```python
>>> idx = pd.IntervalIndex.from_tuples([(1, 5), (3, 7)])
>>> s = pd.Series(range(len(idx)), index=idx)
>>> s
(1, 5]    0
(3, 7]    1
dtype: int64
>>>
>>> s.loc[pd.Interval(3, 5)]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1409, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1827, in _getitem_axis
    return self._get_label(key, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 152, in _get_label
    return self.obj._xs(label, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\generic.py", line 3738, in xs
    loc = self.index.get_loc(key)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexes\interval.py", line 875, in get_loc
    raise KeyError(key)
KeyError: Interval(3, 5, closed='right')
>>>
>>> s.loc[[pd.Interval(3, 5)]]
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1409, in __getitem__
    return self._getitem_axis(maybe_callable, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1816, in _getitem_axis
    return self._getitem_iterable(key, axis=axis)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1118, in _getitem_iterable
    keyarr, indexer = self._get_listlike_indexer(key, axis, raise_missing=False)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexing.py", line 1060, in _get_listlike_indexer
    indexer, keyarr = ax._convert_listlike_indexer(key, kind=self.name)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexes\base.py", line 3239, in _convert_listlike_indexer
    indexer = self._convert_list_indexer(keyarr, kind=kind)
  File "C:\Users\simon\OneDrive\code\pandas-simonjayhawkins\pandas\core\indexes\interval.py", line 626, in _convert_list_indexer
    raise KeyError
KeyError
>>>

```
#### Output of ``pd.show_versions()``

<details>

[paste the output of ``pd.show_versions()`` here below this line]

</details>

### PR DESCRIPTION
**Pull request by phofl**
- [x] closes #27365
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

Without ``tolist`` error would look like ``KeyError: array([6])``

### CODE REVIEW

File: pandas/core/indexes/interval.py [hunk 1]
```diff
@@ -845,7 +845,7 @@ def _convert_list_indexer(self, keyarr):
 
         # we have missing values
         if (locs == -1).any():
-            raise KeyError
+            raise KeyError(keyarr[locs == -1].tolist())
```
Comment (jreback) [inline pandas/core/indexes/interval.py:848]: is this user faing?

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/indexes/interval.py (hunk 1 above)
Comment (phofl) [reply pandas/core/indexes/interval.py:848]: Yep, non existing Keys for IntervalIndex

**Comment by jbrockmendel** [comment]
Tracking this down, it looks like the KeyError raised in Loc._validate_read_indexer has the well-written version of the message, and it looks like it is special-casing Categoricalndex for no apparent reason (CategoricalIndex and IntervalIndex being the only two Index subclasses that do the check this PR is looking at).

Not yet totally sure, but it _may_ be that we should always have this exception be raised in _validate_read_indexer

**Comment by phofl** [comment]
@jbrockmendel I was thinking along the same line. One problem is, that we arrive in ``_get_listlike_indexer`` through ``_getitem_iterable`` where ``raise_missing`` ist set to ``False``. This leads to supressed KeyErrors in some cases when not raising in IntervalIndex implementation. If we could change this to ``True``, we do not need the raise above. But comment says will stop working in the future, so I thought we can not change this now.

**Comment by jbrockmendel** [comment]
> But comment says will stop working in the future, so I thought we can not change this now.

Can you track down when these comments got added?  I think its likely that they are out of date.

**Comment by jbrockmendel** [comment]
@phofl i came to the conclusion that the future-behavior you mentioned was supposed to have already been implemented, so those comments should be removed/updated.

moreover, if we just remove IntervalIndex._convert_list_indexer altogether (and disable it from being called from CategoricalIndex._convert_list_indexer, we end up falling though to raise `raise KeyError(f"None of [{key}] are in the [{axis_name}]")` in `Loc._validate_read_indexer`, like we would for any other Index.

The remaining trouble is that the exception message isn't quite right for IntervalIndex bc it isnt a matter of the key not being present, but the key not being contained in any of the intervals.

**Comment by phofl** [comment]
@jbrockmendel ok great. Will remove the code paths and update this pr with it. Should come to this tomorrow evening

**Comment by phofl** [comment]
Removing

completly does not work out of the box.

```
ser = Series(np.arange(5), IntervalIndex.from_breaks(np.arange(6)))
result = s.loc[[1.5, 2.5, 3.5]]
```

where result has index
``idx=Index([1.5,2.5,3.5])`` instead of intervals.

Additionally

```
df = DataFrame({"A": range(10)})
 s = pd.cut(df.A, 5)
df["B"] = s
df = df.set_index("B")
result = df.loc[[4]]
```

raises ``{KeyError}'a list-indexer must only include values that are in the categories'`` which is probably wrong because we have an interval containing 4?

**Comment by jbrockmendel** [comment]
> completly does not work out of the box.

Thanks for giving it a try.  I may have been overly optimistic about reducing code here.

File: doc/source/whatsnew/v1.2.0.rst [hunk 2]
```diff
@@ -584,6 +584,7 @@ Indexing
 - Bug in :meth:`DataFrame.loc` returned requested key plus missing values when ``loc`` was applied to single level from :class:`MultiIndex` (:issue:`27104`)
 - Bug in indexing on a :class:`Series` or :class:`DataFrame` with a :class:`CategoricalIndex` using a listlike indexer containing NA values (:issue:`37722`)
 - Bug in :meth:`DataFrame.xs` ignored ``droplevel=False`` for columns (:issue:`19056`)
+- Bug in :meth:`Series.__getitem__` and :meth:`DataFrame.__getitem__` returns blank ``KeyError`` without missing keys for :class:`IntervalIndex` (:issue:`27365`)
```
Comment (jbrockmendel) [inline doc/source/whatsnew/v1.2.0.rst:587]: returns -> raising

**Comment by phofl** [comment]
cc @jbrockmendel 
Looked a bit more into this and got a follow up question.

```
df = DataFrame({"A": range(10)})
s = pd.cut(df.A, 5)
df["B"] = s
df = df.set_index("B")
result = df.loc[[4]]
```

This raises because 

```
Int64Index([4], dtype='int64').difference(IntervalIndex([(-0.009, 1.8], (1.8, 3.6], (3.6, 5.4], (5.4, 7.2], (7.2, 9.0]],  closed='right', dtype='interval[float64]')
```

returns
```
Int64Index([4], dtype='int64')
```
 I am not that familiar with interval support in this operation, but this seems at least odd. I would have expected one of the following:

- Raising because incompatible
- Empty return

Is my assumption correct? Then fixing this would bring us one step closer to removing this as you proposed

File: doc/source/whatsnew/v1.2.0.rst (hunk 2 above)
Comment (phofl) [reply doc/source/whatsnew/v1.2.0.rst]: Sorry, changed

**Comment by jbrockmendel** [comment]
> This raises because

This example doesn't raise for me on master.  Do you mean in this branch?

> I am not that familiar with interval support in this operation, but this seems at least odd. I would have expected one of the following:

indexing on a CategoricalIndex is difficult, and I think not entirely consistent (pretty sure there's an issue about this).

IntervalIndex.get_loc/get_indexer behavior is pretty much unconnected to the `.difference` operation (though that would make sense for e.g. 2 `Int64Index`es).  So im not sure i understand the question.

File: pandas/core/indexes/interval.py (hunk 1 above)
Comment (jbrockmendel) [reply pandas/core/indexes/interval.py:829]: @jreback i think this is about ready, only question is if we want to further flesh out the exception message here

**Comment by phofl** [comment]
Sorry, my question was rubbish without context. 

If we remove 

```
     if self.categories._defer_to_indexing:
            indexer = self.categories._convert_list_indexer(keyarr)
            return Index(self.codes).get_indexer_for(indexer)
```

from ``_convert_list_indexer`` in ``category.py`` as discussed above, my example raises. In this case it gets piped through to the code snippet posted above with the ``differences`` return.

**Comment by jbrockmendel** [comment]
> from _convert_list_indexer in category.py as discussed above, my example raises. In this case it gets piped through to the code snippet posted above with the differences return.

yah lets ignore this as orthogonal to this PR.  can revisit later if we get ambitious about code simplification.

**Comment by jreback** [comment]
this looks good, can you merge master @phofl if we want to open an issue about the actual error message ok by me.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @phofl very nice, keep em coming (as you have been!)

**Comment by simonjayhawkins** [comment]
> if we want to open an issue about the actual error message ok by me.

#25996 already exists for key error messages in general.

**Comment by phofl** [comment]
Thanks very much, its fun working with you all

**Pull request merged**